x/subscription/keeper/subscription.go

```diff
@@ -230,80 +236,86 @@ func (k Keeper) CreateSubscription(
 	return err
 }
 
+func (k Keeper) createNewSubscription(ctx sdk.Context, plan *planstypes.Plan, creator, consumer string,
+	block uint64, autoRenewalFlag bool,
+) (types.Subscription, error) {
+	sub := types.Subscription{
+		Creator:       creator,
+		Consumer:      consumer,
+		Block:         block,
+		PlanIndex:     plan.Index,
+		PlanBlock:     plan.Block,
+		DurationTotal: 0,
+		AutoRenewal:   autoRenewalFlag,
+	}
+
+	sub.MonthCuTotal = plan.PlanPolicy.GetTotalCuLimit()
+	sub.MonthCuLeft = plan.PlanPolicy.GetTotalCuLimit()
+	sub.Cluster = types.GetClusterKey(sub)
+
+	// new subscription needs a default project
+	err := k.projectsKeeper.CreateAdminProject(ctx, consumer, *plan)
+	if err != nil {
+		return types.Subscription{}, utils.LavaFormatWarning("failed to create default project", err)
+	}
+
+	return sub, nil
+}
+
+func (k Keeper) upgradeSubscriptionPlan(ctx sdk.Context, duration uint64, sub *types.Subscription, toPlanIndex string, toPlanBlock uint64) error {
+	date := ctx.BlockTime()
+	block := uint64(ctx.BlockHeight())
+
+	k.addCuTrackerTimerForSubscription(ctx, block, sub)
+
+	if sub.DurationLeft == 0 {
+		// Subscription was already expired. Can't upgrade.
+		k.handleZeroDurationLeftForSubscription(ctx, block, date, sub)
+		return utils.LavaFormatError("Trying to upgrade subscription, but subscription was already expired", nil,
+			utils.LogAttr("consumer", sub.Consumer))
+	}
+
+	// The "old" subscription's duration is now expired
+	// If called from CreateSubscription, the duration will reset to the duration bought
+	sub.DurationLeft = 0
+	sub.Cluster = types.GetClusterKey(*sub)
+
+	nextEpoch, err := k.epochstorageKeeper.GetNextEpoch(ctx, block)
+	if err != nil {
+		return utils.LavaFormatError("Trying to upgrade subscription, but got an error while trying to get next epoch", err,
+			utils.LogAttr("consumer", sub.Consumer))
+	}
+
+	sub.PlanIndex = toPlanIndex
+	sub.PlanBlock = toPlanBlock
+
+	k.resetSubscriptionDetailsAndAppendEntry(ctx, sub, nextEpoch, date)
+
```

Comment:
also we need to remove the previous timer of the end of the month since we did it here (the advanceMonth timer):
if we bought on 1/1/23 we will have a timer for the nextmonth at 1/2/23
if we upgrade in the middle of the month, lets say 15/1/23, we will do the upgrade and open a timer for the 15/2/23.
BUT we did not remove the 1/2/23 timer

Reply:
How do I remove a time from TimerStore? :sweat_smile:

Reply:
Fixed in 117d475919b5c8b9962ea4b52917ee34aeabeec5